src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs

```diff
@@ -15,7 +15,105 @@
 namespace Microsoft.PowerFx.Functions
 {
     internal static partial class Library
-    {
+    {
+        public static async ValueTask<FormulaValue> Collect(EvalVisitor runner, EvalVisitorContext context, IRContext irContext, FormulaValue[] args)
+        {
+            FormulaValue arg0;
+            var argc = args.Length;
+            var returnIsTable = args.Length > 2;
+
+            // Need to check if the Lazy first argument has been evaluated since it may have already been
+            // evaluated in the ClearCollect case.
+            if (args[0] is LambdaFormulaValue arg0lazy)
+            {
+                arg0 = await arg0lazy.EvalAsync().ConfigureAwait(false);
+            }
+            else
+            {
+                arg0 = args[0];
+            }
+
+            if (arg0 is BlankValue)
+            {
+                return arg0;
+            }
+            else if (arg0 is ErrorValue)
+            {
+                return arg0;
+            }
+
+            if (arg0 is not TableValue)
+            {
+                return CommonErrors.RuntimeTypeMismatch(IRContext.NotInSource(arg0.Type));
+            }
+
+            var tableValue = arg0 as TableValue;
+
+            List<DValue<RecordValue>> resultRows = new List<DValue<RecordValue>>();
+
+            for (int i = 1; i < argc; i++)
+            {
+                runner.CheckCancel();
+
+                var arg = args[i];
+
+                if (arg is TableValue argTableValue)
+                {
+                    returnIsTable = true;
+
+                    foreach (DValue<RecordValue> row in argTableValue.Rows)
+                    {
+                        if (row.IsBlank)
+                        {
+                            continue;
+                        }
+                        else if (row.IsError)
+                        {
+                            return row.Error;
+                        }
+                        else
+                        {
+                            var recordValueCopy = FormulaValue.NewRecordFromFields(row.Value.Fields);
+                            resultRows.Add(await tableValue.AppendAsync(recordValueCopy, runner.CancellationToken).ConfigureAwait(false));
+                        }
+                    }
+                }
+                else if (arg is RecordValue recordValue)
+                {
+                    var recordValueCopy = FormulaValue.NewRecordFromFields(recordValue.Fields);
+                    resultRows.Add(await tableValue.AppendAsync(recordValueCopy, runner.CancellationToken).ConfigureAwait(false));
+                }
+                else if (arg is ErrorValue)
+                {
+                    return arg;
+                }
+                else if (arg is BlankValue && !tableValue.Type._type.IsSingleColumnTable)
+                {
+                    continue;
+                }
+                else

```

Comment:
>else [](http://example.com/codeflow?start=16&length=4)

Is this the case for primitive values (scalar Collect)? If so, please add a comment